### ISSUE TITLE
Exceptions related to socket timeout with MySQL driver

### ISSUE THREAD

**Issue by redshiftetl** [brettwooldridge/HikariCP#310]
This issue seems similar to #236, which I believe was fixed in version 2.3.1 - I'm on version 2.3.6. I'm seeing a lot of two possibly related exceptions in my production environment:

````
2015-04-17 07:16:26,812 | ERROR | Hikari JDBC Timeout Executor | o.a.z.s.NIOServerCnxnFactory | Thread Thread[Hikari JDBC Timeout Executor,5,main] died
java.lang.NullPointerException: null
	at com.mysql.jdbc.MysqlIO.setSocketTimeout(MysqlIO.java:4838) ~[mysql.mysql-connector-java-5.1.35.jar:5.1.35]
	at com.mysql.jdbc.ConnectionImpl$12.run(ConnectionImpl.java:5544) ~[mysql.mysql-connector-java-5.1.35.jar:5.1.35]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) ~[na:1.8.0_40-internal]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) ~[na:1.8.0_40-internal]
	at java.lang.Thread.run(Thread.java:745) ~[na:1.8.0_40-internal]
````

and

````
2015-04-17 07:16:26,907 | ERROR | Hikari JDBC Timeout Executor | o.a.z.s.NIOServerCnxnFactory | Thread Thread[Hikari JDBC Timeout Executor,5,main] died
java.net.SocketException: Socket is closed
	at java.net.Socket.setSoTimeout(Socket.java:1137) ~[na:1.8.0_40-internal]
	at com.mysql.jdbc.MysqlIO.setSocketTimeout(MysqlIO.java:4838) ~[mysql.mysql-connector-java-5.1.35.jar:5.1.35]
	... 4 common frames omitted
Wrapped by: java.sql.SQLException: Invalid socket timeout value or state
	at com.mysql.jdbc.SQLError.createSQLException(SQLError.java:998) ~[mysql.mysql-connector-java-5.1.35.jar:5.1.35]
	at com.mysql.jdbc.SQLError.createSQLException(SQLError.java:937) ~[mysql.mysql-connector-java-5.1.35.jar:5.1.35]
	at com.mysql.jdbc.SQLError.createSQLException(SQLError.java:926) ~[mysql.mysql-connector-java-5.1.35.jar:5.1.35]
	at com.mysql.jdbc.SQLError.createSQLException(SQLError.java:872) ~[mysql.mysql-connector-java-5.1.35.jar:5.1.35]
	at com.mysql.jdbc.MysqlIO.setSocketTimeout(MysqlIO.java:4840) ~[mysql.mysql-connector-java-5.1.35.jar:5.1.35]
	at com.mysql.jdbc.ConnectionImpl$12.run(ConnectionImpl.java:5544) ~[mysql.mysql-connector-java-5.1.35.jar:5.1.35]
	... 3 common frames omitted
Wrapped by: java.lang.RuntimeException: java.sql.SQLException: Invalid socket timeout value or state
	at com.mysql.jdbc.ConnectionImpl$12.run(ConnectionImpl.java:5546) ~[mysql.mysql-connector-java-5.1.35.jar:5.1.35]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) ~[na:1.8.0_40-internal]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) ~[na:1.8.0_40-internal]
	at java.lang.Thread.run(Thread.java:745) ~[na:1.8.0_40-internal]
````

Any other information that would be useful? Thanks for your help.

**Comment by brettwooldridge** [comment]
@redshiftetl I believe this was fixed by this recent check-in 20f47c3 on the ``2.3.x`` branch.  It is clear that the original code did not fix the issue for users using ``jdbcUrl`` or ``driverClassName`` based configuration.  It only fixed it in the case of using ``dataSourceClassName``.  The fix will be in release 2.3.7.

**Comment by redshiftetl** [comment]
Got it - thanks!
